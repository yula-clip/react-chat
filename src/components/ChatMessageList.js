import React from 'react';
import { withRouter } from 'react-router-dom';
import { withStyles } from '@material-ui/core/styles';
import ChatMessage from './ChatMessage';
import Typography from '@material-ui/core/Typography';
import Paper from '@material-ui/core/Paper';

const styles = theme => ({
  messagesWrapper: {
    overflowX: 'scroll',
    height: '100%',
    width: '100%',
    paddingTop: theme.spacing.unit * 3,
    paddingBottom: '120px',
  },
  paper: {
    padding: theme.spacing.unit * 3,
  }
});

class ChatMessageList extends React.Component {
  componentDidMount() {
    this.scrollDownHistory();
  }

  componentDidUpdate() {
    this.scrollDownHistory();
  }

  scrollDownHistory() {
    const messagesWrapper = this.refs.messagesWrapper;
    if (messagesWrapper) {
      messagesWrapper.scrollTop = messagesWrapper.scrollHeight;
    }
  }

  render() {
    const { classes, messages, match, activeUser } = this.props;
    if (!match.params.chatId) {
      return (
        <Paper className={classes.paper}>
          <Typography variant="display1" gutterBottom>
            Start messagingâ€¦
          </Typography>
          <Typography variant="body1" gutterBottom>
            Use <strong>Global</strong> to explore communities around here.
          </Typography>
          <Typography variant="body1" gutterBottom>
            Use <strong>Recents</strong> to see your recent conversations.
          </Typography>
        </Paper>
      );
    }
    return messages && messages.length ? (
      <div className={classes.messagesWrapper} ref="messagesWrapper">
        {messages.map((message, index) => (
          <ChatMessage key={index} activeUser={activeUser} {...message} />
        ))}
      </div>
    ) : (
        <Typography variant="display1">
          There is no messages yet...
        </Typography>
        );
  }
}

export default withRouter(withStyles(styles)(ChatMessageList));

