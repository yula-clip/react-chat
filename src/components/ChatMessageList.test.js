/* eslint-env jest */
import React from "react";
import ReactDOM from "react-dom";
import { MemoryRouter, Route } from "react-router-dom";
import renderer from "react-test-renderer";
import ChatMessageList from "./ChatMessageList";

jest.mock("moment", () => () => ({ fromNow: () => "2 days ago" }));

const mockProps = {
  messages: [
    {
      _id: "123foo321",
      chatId: "12345",
      content: "Hello, World!",
      sender: {
        _id: "12345",
        username: "me"
      },
      createdAt: "2018-03-16T10:53:23.200Z"
    },
    {
      _id: "321bar123",
      chatId: "12345",
      content: "Hello, React!",
      sender: {
        _id: "54321",
        username: "someone"
      },
      createdAt: "2018-03-16T10:53:23.200Z"
    }
  ],
  activeUser: {
    _id: "12345",
    username: "me",
    isMember: true,
    isCreator: true,
    isChatMember: true
  }
};

describe("<ChatMessageList />", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter initialEntries={["/chat"]}>
        <Route
          path="/chat/:chatId?"
          render={props => <ChatMessageList {...mockProps} {...props} />}
        />
      </MemoryRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders paper", () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/chat"]}>
          <Route
            path="/chat/:chatId?"
            render={props => <ChatMessageList {...mockProps} {...props} />}
          />
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders messages", () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/chat/12345"]}>
          <Route
            path="/chat/:chatId?"
            render={props => <ChatMessageList {...mockProps} {...props} />}
          />
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });

  it("renders without messages", () => {
    const tree = renderer
      .create(
        <MemoryRouter initialEntries={["/chat/12345"]}>
          <Route
            path="/chat/:chatId?"
            render={props => (
              <ChatMessageList {...mockProps} {...props} messages={[]} />
            )}
          />
        </MemoryRouter>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
