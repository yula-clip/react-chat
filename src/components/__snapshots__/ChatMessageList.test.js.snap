// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ChatMessageList /> renders messages 1`] = `
<div
  className="ChatMessageList-messagesWrapper-65"
>
  <div
    className="ChatMessage-messageWrapper-132 ChatMessage-messageWrapperFromMe-129"
  >
    <div
      className="MuiPaper-root-67 MuiPaper-elevation2-71 MuiPaper-rounded-68 ChatMessage-message-130 ChatMessage-messageFromMe-131"
    >
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
        style={
          Object {
            "color": "#f44336",
          }
        }
      >
        me
      </span>
      <p
        className="MuiTypography-root-94 MuiTypography-body1-103"
      >
        Hello, World!
      </p>
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
      >
        2 days ago
      </span>
    </div>
    <div
      className="MuiAvatar-root-135 MuiAvatar-colorDefault-136"
      style={
        Object {
          "backgroundColor": "#f44336",
        }
      }
    >
      M
    </div>
  </div>
  <div
    className="ChatMessage-messageWrapper-132"
  >
    <div
      className="MuiAvatar-root-135 MuiAvatar-colorDefault-136"
      style={
        Object {
          "backgroundColor": "#f44336",
        }
      }
    >
      S
    </div>
    <div
      className="MuiPaper-root-67 MuiPaper-elevation2-71 MuiPaper-rounded-68 ChatMessage-message-130"
    >
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
        style={
          Object {
            "color": "#f44336",
          }
        }
      >
        someone
      </span>
      <p
        className="MuiTypography-root-94 MuiTypography-body1-103"
      >
        Hello, React!
      </p>
      <span
        className="MuiTypography-root-94 MuiTypography-caption-104"
      >
        2 days ago
      </span>
    </div>
  </div>
</div>
`;

exports[`<ChatMessageList /> renders paper 1`] = `
<div
  className="MuiPaper-root-67 MuiPaper-elevation2-71 MuiPaper-rounded-68 ChatMessageList-paper-66"
>
  <h4
    className="MuiTypography-root-94 MuiTypography-h4-109 MuiTypography-gutterBottom-121"
  >
    Start messagingâ€¦
  </h4>
  <p
    className="MuiTypography-root-94 MuiTypography-body1-103 MuiTypography-gutterBottom-121"
  >
    Use 
    <strong>
      Global
    </strong>
     to explore communities around here.
  </p>
  <p
    className="MuiTypography-root-94 MuiTypography-body1-103 MuiTypography-gutterBottom-121"
  >
    Use 
    <strong>
      Recents
    </strong>
     to see your recent conversations.
  </p>
</div>
`;

exports[`<ChatMessageList /> renders without messages 1`] = `
<h4
  className="MuiTypography-root-94 MuiTypography-h4-109"
>
  There is no messages yet...
</h4>
`;
